core/src/main/scala/org/apache/spark/sql/execution/columnar/encoding/ColumnEncoding.scala

```diff
@@ -0,0 +1,346 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you
+ * may not use this file except in compliance with the License. You
+ * may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied. See the License for the specific language governing
+ * permissions and limitations under the License. See accompanying
+ * LICENSE file.
+ */
+package org.apache.spark.sql.execution.columnar.encoding
+
+import java.lang.reflect.Field
+import java.nio.ByteOrder
+
+import org.apache.spark.sql.catalyst.expressions.{Attribute, UnsafeArrayData, UnsafeMapData, UnsafeRow}
+import org.apache.spark.sql.types._
+import org.apache.spark.unsafe.Platform
+import org.apache.spark.unsafe.types.{CalendarInterval, UTF8String}
+import org.apache.spark.util.collection.BitSet
+
+private[columnar] abstract class ColumnEncoding {
+
+  private[columnar] final var cursor = 0
+
+  def typeId: Int
+
+  def supports(dataType: DataType): Boolean
+
+  def initializeDecoding(columnBytes: Array[Byte], field: Attribute): Unit
+
+  def readNull(columnBytes: Array[Byte], ordinal: Int): Boolean
+
+  def readBoolean(columnBytes: Array[Byte]): Boolean
+
+  def nextBoolean(columnBytes: Array[Byte]): Unit
+
+  def readByte(columnBytes: Array[Byte]): Byte
+
+  def nextByte(columnBytes: Array[Byte]): Unit
+
+  def readShort(columnBytes: Array[Byte]): Short
+
+  def nextShort(columnBytes: Array[Byte]): Unit
+
+  def readInt(columnBytes: Array[Byte]): Int
+
+  def nextInt(columnBytes: Array[Byte]): Unit
+
+  def readLong(columnBytes: Array[Byte]): Long
+
+  def nextLong(columnBytes: Array[Byte]): Unit
+
+  def readFloat(columnBytes: Array[Byte]): Float
+
+  def nextFloat(columnBytes: Array[Byte]): Unit
+
+  def readDouble(columnBytes: Array[Byte]): Double
+
+  def nextDouble(columnBytes: Array[Byte]): Unit
+
+  def readDecimal(columnBytes: Array[Byte], precision: Int,
+      scale: Int): Decimal
+
+  def nextDecimal(columnBytes: Array[Byte], precision: Int): Unit
+
+  def readUTF8String(columnBytes: Array[Byte]): UTF8String
+
+  def nextUTF8String(columnBytes: Array[Byte]): Unit
+
+  def readBinary(columnBytes: Array[Byte]): Array[Byte]
+
+  def nextBinary(columnBytes: Array[Byte]): Unit
+
+  def readInterval(columnBytes: Array[Byte]): CalendarInterval
+
+  def nextInterval(columnBytes: Array[Byte]): Unit
+
+  def readArray(columnBytes: Array[Byte]): UnsafeArrayData
+
+  def readMap(columnBytes: Array[Byte]): UnsafeMapData
+
+  def readStruct(columnBytes: Array[Byte], numFields: Int): UnsafeRow
+
+  def initializeEncoding(dataType: DataType, batchSize: Int): Array[Byte]
+
+  def writeBoolean(columnBytes: Array[Byte], value: Boolean): Unit
+
+  def writeByte(columnBytes: Array[Byte], value: Byte): Unit
```

Comment:
What are these write functions for? I thought we just have to read the encoding, we would still use Spark code to encode.